### DISCUSSION TITLE
Add option to toggle entity for specific time

### DISCUSSION THREAD

**Discussion by Schnabulation** [home-assistant/frontend#21677] [Dashboard]
![HA_mockup Kopie](https://github.com/user-attachments/assets/f2e26d07-f5a9-4a6c-8f80-b095f7081e47)

**Idea:** Sometimes the user might want to turn something on and have it automatically turn off after a set amount of time. For that to work the user has to write a seperate script or automation for this specific entity.

**Solution:** Every entity of type switch should populate a couple of buttons in the HA frontend whereas the user can toggle the actual state of the switch and have it revert after the selected time. Home Assistant could provide a set of regular time intervals (ex. 30 mins, 60 mins, 120 mins) but leave the option to insert a free value.

**Open to discussion:**

- What happens, when the state of the entity changes in the meantime?
- Is it reboot persistent?
- Can the user see the timer and interact with it?

**Comment by rlinick** [comment]
Would love this and use it frequently if it were implemented.

**Comment by hammadbinarif** [comment]
I am looking for exactly the same feature. I can do a script for it but it would be wonderful if some card can do it for all my light entities just out of the box

**Comment by RoganDawes** [comment]
It seems to me like this is a UI for creating very simple, temporary automations that delete themselves when they are done. Perhaps framing it like that would help?

**Reply by Schnabulation** [reply]
While I do understand what you mean, I'm not really sure, if this is the correct approach. Creating temporary automations would sure be a method to solve this but it would be cumbersome in my opinion. Having a dedicated button or menu-item auto-created for each entity would be the way to go. Furthermore using this button should not rely on automations, but be fully backend-supported. Otherwise you would have the whole overhead for creating a automation (entity_id, naming, etc.). I'd say this is something that should be done in the background by HA core.

**Reply by RoganDawes** [reply]
All I mean is that the backend infrastructure for managing such a feature already exists, in the automations. Whether these become a different category of automation, created and managed by HA, and therefore less visible to the user via the conventional Automations UI is one way of keeping them separate from the user's own custom automations.

Plus, HA already has a way of identifying automations that affect a particular device/entity, which could potentially be useful when constructing a frontend interface. e.g. when presenting a switch entity (which I think this request is mostly targeted at), query for any related automations and filter for those matching this new category), so that the UI can be appropriately decorated (with remaining time, or whatever).

**Reply by RoganDawes** [reply]
That also (partially) answers the question of whether these can (not should) persist across a reboot. If saved as an automation with a precalculated trigger time rather than a relative delay, the automation absolutely could persist across a reboot. I guess it would need to have an "on_boot" trigger to be cleaned up if the trigger time has already passed when the reboot happens, though. And the automation could check that the entity is in the expected state on boot, that it has not been changed externally while HA was rebooting.

And I'm pretty sure the automation can be constructed to trigger on any state change of the related entity before the trigger time and deactivate/delete itself, perhaps with a notification.

**Comment by RoganDawes** [comment]
Just one addition. I think there are two possible use cases for such temporary automations:
1. Toggle switch for X duration,
2. Wait X duration and then toggle the switch.

Real world examples:
I want to disable an automation for the next 24 hours.
I have manually turned on a light, and want it to turn off again in 15 minutes.

Edit: I think going with "toggle" is a bad idea, as it may end up undoing a manual change in between. Better to explicitly specify `switch.turn_on` or `switch.turn_off` the entity

**Reply by Schnabulation** [reply]
Your suggestion regarding "wait X then toggle" is great! That would be another really useful addition.

My take on "toggle" instead of "turn on / off" was that it would be simpler to display on the frontend. However we should clarify what happens when the switch gets manually changed while the system is running. I personally would say the system gets stopped.

**Reply by RoganDawes** [reply]
Perhaps this could be an additional item under the "three dots" menu for switches? -> Delayed Operation -> Wait then toggle / toggle for ...

**Reply by RoganDawes** [reply]
Another thought might be to have just a single "Timed operation" menu item. It brings up a dialog where you can hit a checkbox for "Toggle now", then choose the time for the next operation. Can default to a period delay, say 60 minutes/1 hour, but could be switchable to a precise time. Wording of switching operations should take into account the current state of the switch, i.e. "Switch off now", "Wait 60 minutes", "Switch on". If the initial checkbox for "Switch off now" is not checked, the subsequent operation should be the inverse of the current state.

**Comment by karwosts** [comment]
While I can't really imagine this coming to the main HA UI, If you wanted to implement this for yourself you could make a script with an entity selector and a time selector as fields, and that could turn on the entity, wait for the delay, and turn it off again. 

Then you could just launch this script from your UI where it will prompt you to fill in the fields. So not much more steps than would take here.

**Reply by Schnabulation** [reply]
Thank you for the suggestion. I have actually already created this in the meantime:

![grafik](https://github.com/user-attachments/assets/1c19d79f-8650-4c0b-a4b8-f99487fa2035)

```
alias: "Frontend: Toggle for set time"
description: Toggles an entity for a given duration
sequence:
  - action: homeassistant.toggle
    target:
      entity_id: "{{ entity }}"
  - delay: "{{ time }}"
  - action: homeassistant.toggle
    target:
      entity_id: "{{ entity }}"
fields:
  entity:
    selector:
      entity:
        filter:
          domain:
            - light
            - switch
            - fan
    name: Schalter auswählen
    required: true
  time:
    selector:
      duration: {}
    name: Schaltdauer auswählen
    required: true
mode: parallel
max: 10
```

**Reply by fukuoka19** [reply]
I am new to HA and terrible at programming.  I like what you have proposed, but was hoping you could give me a little more help.  Where/how did you create the card to insert the programing above?

**Reply by HenrikHills** [reply]
I think its just a script, and the ‘card’ you see is the more-info tab of that script

**Reply by Schnabulation** [reply]
You sir, are correct!
